lib/Frontend/CompilerInvocation.cpp

```diff
@@ -973,6 +645,225 @@ StringRef FrontendArgsToOptionsConverter::determineFallbackModuleName() const {
   return llvm::sys::path::stem(nameToStem);
 }
 
+bool FrontendArgsToOptionsConverter::setOutputFilenames() {
+  const std::vector<std::string> unprocessedOutputFlienames =
+  getUnprocessedOutputFilenames();
+  
+  if (!unprocessedOutputFlienames.empty() &&
+      !llvm::sys::fs::is_directory(unprocessedOutputFlienames.back())) {
+    Opts.OutputFilenames = unprocessedOutputFlienames;
+    return false;
+  }
+  if (Opts.Inputs.isReadingFromStdin() && unprocessedOutputFlienames.empty()) {
+    Opts.OutputFilenames = unprocessedOutputFlienames;
+    return false;
+  }
+  
+  // No output filename was specified, or an output directory was specified.
+  // Determine the correct output filename.
+  
+  // Note: this should typically only be used when invoking the frontend
+  // directly, as the driver will always pass -o with an appropriate filename
+  // if output is required for the requested action.
+  
+  assert(Opts.actionProducesOutputFromFrontend() ||
+         Opts.OutputFilenames.empty());
+  assert(Opts.actionProducesOutputFromFrontend() ||
+         !Opts.actionOutputsToStdout());
+  if (unprocessedOutputFlienames.empty() &&
+      (Opts.Inputs.isReadingFromStdin() || Opts.actionOutputsToStdout())) {
+    Opts.setOutputFilenameToStdout();
+    return false;
+  }
+  return deriveOutputFilenamesFromInputsAndSuffix();
+}
+
+bool FrontendArgsToOptionsConverter::deriveOutputFilenamesFromInputsAndSuffix() {
+  const std::vector<std::string> unprocessedOutputFlienames =
+  getUnprocessedOutputFilenames();
+  {
+    const unsigned commandLineOutputFilenameCount =
```

Comment:
Not necessarily "command line", since this includes file lists.

Reply:
Good catch! Renamed commandLineOutputFilenameCount to unprocessedOutputFilenameCount.